pkg/components/ebpf/common/kafkaparser/common.go

```diff
@@ -0,0 +1,307 @@
+}
+
+func validateKafkaRequestHeader(header *KafkaRequestHeader) error {
+	if header.MessageSize < int32(MinKafkaRequestLen) || header.APIVersion < 0 {
+		return errors.New("invalid Kafka request header: size or version is negative")
+	}
+
+	if header.MessageSize > KafkaMaxPayloadLen {
+		return errors.New("invalid Kafka request header: message size exceeds maximum payload length")
+	}
+
+	switch header.APIKey {
+	case APIKeyFetch:
+		if header.APIVersion > 17 { // latest: Fetch Request (Version: 17)
+			return errors.New("invalid Kafka request header: unsupported API key version for Fetch")
+		}
+	case APIKeyProduce:
+		if header.APIVersion > 12 { // latest: Produce Request (Version: 12)
+			return errors.New("invalid Kafka request header: unsupported API key version for Produce")
+		}
+	case APIKeyMetadata:
+		if header.APIVersion < 10 || header.APIVersion > 13 { // latest: Metadata Request (Version: 13), only versions 10-13 contain topic_id which we are interested in
+			return errors.New("invalid Kafka request header: unsupported API key version for Metadata")
+		}
+	default:
+		return errors.New("invalid Kafka request header: unsupported API key")
+	}
+	if header.CorrelationID < 0 {
+		return errors.New("invalid Kafka request header: correlation ID is negative")
+	}
+	return nil
+}
+
+func validateKafkaResponseHeader(header *KafkaResponseHeader, requestHeader *KafkaRequestHeader) error {
+	if header.MessageSize < int32(MinKafkaRequestLen) {
+		return errors.New("invalid Kafka response header: size is negative")
+	}
+
+	if header.MessageSize > KafkaMaxPayloadLen {
+		return errors.New("invalid Kafka response header: message size exceeds maximum payload length")
+	}
+
+	if header.CorrelationID < 0 {
+		return errors.New("invalid Kafka request header: correlation ID is negative")
+	}
+	if header.CorrelationID != requestHeader.CorrelationID {
+		return errors.New("invalid Kafka response header: correlation ID does not match request header")
+	}
+	return nil
+}
+
+// isFlexible checks for each API key if the version is flexible.
+// a flexible version uses a dynamic size for arrays and strings
+func isFlexible(header *KafkaRequestHeader) bool {
+	switch header.APIKey {
+	// https://github.com/apache/kafka/blob/9983331d917fe8f57c37c88f0749b757e5af0c87/clients/src/main/resources/common/message/ProduceRequest.json#L51
+	case APIKeyProduce:
+		return header.APIVersion >= 9
+	// https://github.com/apache/kafka/blob/9983331d917fe8f57c37c88f0749b757e5af0c87/clients/src/main/resources/common/message/FetchRequest.json#L62C4-L62C20
+	case APIKeyFetch:
+		return header.APIVersion >= 12
+	// https://github.com/apache/kafka/blob/9983331d917fe8f57c37c88f0749b757e5af0c87/clients/src/main/resources/common/message/MetadataRequest.json#L22
+	case APIKeyMetadata:
+		return header.APIVersion >= 9
+	default:
+		return false
+	}
+}
+
+func readArrayLength(pkt []byte, header *KafkaRequestHeader, offset Offset) (int, Offset, error) {
+	if isFlexible(header) {
```

Comment:
do we need remaining length check on the packet here?

Reply:
done